### PR TITLE
Add Pressure-based rate limiter

### PR DESCRIPTION
**Pull request by rijkvanzanten**
When you're overloading the system, the API will keep accepting requests, even though the server might not be able to handle them. This PR introduces a new pressure-based rate limiter which will continuously monitor the current event loop and memory usage, and error out requests with a 503 early when the system is overloaded. 

I've opted to have it enabled by default, with some very near-catastrophe limits.

### Other changes:

- I was getting a little annoyed at some of the reliability issues of `@ngneat/falso`, so opted to start a new `@directus/random` package. Doesn't really do much yet, but we'll add more random-functions over time
- Added a new `defaults` util function to `@directus/utils`. Effectively a small wrapper around basic object assignment, but includes some TypeScript magic to make typing options with defaults easier

### CODE REVIEW

File: packages/pressure/src/monitor.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+			maxMemoryRss: false,
+			maxEventLoopDelay: false,
+			maxEventLoopUtilization: false,
+		});
+
+		this.histogram = monitorEventLoopDelay({ resolution: this.options.resolution });
+		this.histogram.enable();
+
+		this.updateUsage = this.updateUsage.bind(this);
+		this.timeout = setTimeout(this.updateUsage, this.options.sampleInterval);
```
Comment (br41nslug) [inline packages/pressure/src/monitor.ts:39]:
```suggestion
		this.timeout = setTimeout(this.updateUsage, Number(this.options.sampleInterval));
```
I think we should also do some more sanity checking beyond just casting it as a number because `setTimeout` accepts strings and infinity's without error but silently falls back to zero or one as timeout 😬 meaning the `updateUsage` function becomes an infinite loop firing every millisecond.

**Review by br41nslug — COMMENTED** [review]
I've been trying to get this pressure rate-limiter to trigger using `k6` but have not managed to do so yet. Are there specific configurations or limits i can set that make it easier for the rate limiter to kick in?

File: packages/random/src/random-int-between.ts
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply packages/random/src/random-int-between.ts:1]: > Having a new package for this one basic function seems a tad like overkill to me in a PR which also adds 2 other packages.

Very overkill! 

> Should this function not be an utility in either @directus/shared or the new @directus/utils?

No!

The goal of `@directus/random` is to replace `@ngneat/falso`. I'd like to keep it a separate package, as they're exclusively intended to be used as a dev-dependency, and are explicitly not built with performance or prod use in mind (it's a testing util). There's no penalty to having additional smaller packages, and the code organization is a big benefit 👍🏻

File: api/src/env.ts [hunk 2]
```diff
@@ -284,6 +284,14 @@ const defaults: Record<string, any> = {
...
 	FLOWS_EXEC_ALLOWED_MODULES: false,
 	FLOWS_ENV_ALLOW_LIST: false,
+
+	PRESSURE_LIMITER_ENABLED: true,
+	PRESSURE_LIMITER_SAMPLE_INTERVAL: 250,
+	PRESSURE_LIMITER_MAX_EVENT_LOOP_UTILIZATION: 0.99,
+	PRESSURE_LIMITER_MAX_EVENT_LOOP_DELAY: 500,
+	PRESSURE_LIMITER_MAX_MEMORY_RSS: false,
+	PRESSURE_LIMITER_MAX_MEMORY_HEAP_USED: false,
+	PRESSURE_LIMITER_RETRY_AFTER: '1',
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/env.ts:294]: The `retry-after` controls the `Retry-After` http header that's given to the client. It has nothing to do with how often the pressure monitor itself checks for the current pressure state 👍🏻

File: packages/utils/src/defaults.ts
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply packages/utils/src/defaults.ts:1]: This package already exists 😄 

The longer term goal is to remove `@directus/shared` in favor of smaller packages: `@directus/types`, `@directus/utils`, `@directus/components` etc

[the remark this answers is not in the document]
Comment (br41nslug) [reply packages/utils/src/defaults.ts:1]: Lol really? wow, my bad 😂

File: api/src/env.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (br41nslug) [reply api/src/env.ts:294]: Then does this label need a default value? 🤔 I can understand sending this header in some situations for client compatibility but as the retry-after value has no real connection to when retrying would actually be valuable perhaps it should not have a default value.

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply api/src/env.ts:300]: Good point!

File: api/src/env.ts [hunk 3]
```diff
@@ -291,6 +291,14 @@ const defaults: Record<string, any> = {
 
 	FLOWS_EXEC_ALLOWED_MODULES: false,
 	FLOWS_ENV_ALLOW_LIST: false,
+
+	PRESSURE_LIMITER_ENABLED: true,
```
Comment (br41nslug) [inline api/src/env.ts:295]: I personally think this should be an opt-in feature instead of opt-out as misconfiguring this can impact the availability and user experience of Directus instances.

**Review by br41nslug — COMMENTED** [review]
The limiter is working as expected in my tests. However some documentation on what each configuration value means and how it impacts your instance would be much appreciated.

**Review by jaads — CHANGES_REQUESTED** [review]
As stated above and as @br41nslug also mentioned, it would be nice to have the new configuration options added in the docs , right? Now that the docs are in the mono repo it might make sense to add those in this PR 🤔

**Comment by br41nslug** [comment]
For some reason the mocking of `node:perf_hooks` and `node:timers` broke. Presumably the cause is the ESM update.

**Review by jaads — APPROVED** [review]
Code and tests are looking good and the docs have been added  👍 
I also made a load test with K6 and the server indeed returns a proper error message when the maximum event loop utilization or the heap limit is exceeded 👍

File: docs/self-hosted/config-options.md [hunk 4]
```diff
@@ -446,6 +446,20 @@ RATE_LIMITER_REDIS_PORT=6379
...
+|-----------------------------------------------|---------------------------------------------------------------------|---------------|
+| `PRESSURE_LIMITER_ENABLED`                    | Whether or not to enable pressure-based rate limiting on the API.   | `true`        |
+| `PRESSURE_LIMITER_SAMPLE_INTERVAL`            | The time window for measuring pressure in ms.                       | `250`         |
+| `PRESSURE_LIMITER_MAX_EVENT_LOOP_UTILIZATION` | The maximum allowed utilization where `1` is 100% loop utilization. | `0.99`        |
+| `PRESSURE_LIMITER_MAX_EVENT_LOOP_DELAY`       | The maximum amount of time the current loop can be delayed in ms.   | `500`         |
+| `PRESSURE_LIMITER_MAX_MEMORY_RSS`             | The maximum allowed memory Resident Set Size (RSS) in bytes.        | `false`       |
+| `PRESSURE_LIMITER_MAX_MEMORY_HEAP_USED`       | The maximum allowed heap usage in bytes.                            | `false`       |
+| `PRESSURE_LIMITER_RETRY_AFTER`                | Sets the `Retry-After` header when the rate limiter is triggered.   | `false`       |
```
Comment (paescuj) [inline docs/self-hosted/config-options.md:461]: What's the reason for having `Retry-After` header disabled by default? Security concerns?

File: packages/pressure/src/monitor.test.ts [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		expect(monitor['updateMemoryUsage']).toHaveBeenCalledOnce();
+	});
+
+	test('Calls updateEventLoopUsage', () => {
+		expect(monitor['updateEventLoopUsage']).toHaveBeenCalledOnce();
+	});
+
+	test('Refreshes the timer', () => {
+		expect(mockTimer.refresh).toHaveBeenCalledOnce();
+	});
```
Comment (paescuj) [inline packages/pressure/src/monitor.test.ts:142]: (Could be one test)

**Review by paescuj — APPROVED** [review]
Some marginal notes and question, but LGTM 🥇

File: docs/self-hosted/config-options.md (hunk 4 above)
Comment (rijkvanzanten) [reply docs/self-hosted/config-options.md:461]: Tim's concerns 🙂 

https://github.com/directus/directus/pull/17873#discussion_r1146295732

The TL;DR is: the retry-header isn't actually connected to anything. It's an arbitrary number that just allows the server to breath for a second before the client retries, assuming that the client checks for that header.

File: .changeset/slimy-zebras-jam.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+"@directus/app": patch
+"@directus/api": patch
```
Comment (rijkvanzanten) [inline .changeset/slimy-zebras-jam.md:3]:
```suggestion
"@directus/api": minor
```

File: packages/pressure/package.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
+{
+	"name": "@directus/pressure",
+	"version": "9.23.1",
+	"type": "module",
+	"scripts": {
+		"build": "tsc --project tsconfig.prod.json",
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply packages/pressure/package.json:6]: Agreed!

File: docs/self-hosted/config-options.md (hunk 4 above)
Comment (paescuj) [reply docs/self-hosted/config-options.md:461]: All clear, thanks!
